### PR TITLE
Fix Rooted FS tests on Windows

### PR DESCRIPTION
**Pull request by pyckle**

### CODE REVIEW

File: sshd-common/src/main/java/org/apache/sshd/common/util/io/IoUtils.java [hunk 1]
```diff
@@ -426,29 +426,40 @@ public static Boolean checkFileExistsAnySymlinks(Path path, boolean neverFollowS
...
         String firstName = path.getName(0).toString();
         String[] names = new String[partsToExtract - 1];
         for (int j = 1; j < partsToExtract; j++) {
             names[j - 1] = path.getName(j).toString();
         }
         Path checkForSymLink = path.getFileSystem().getPath(firstName, names);
+        // the root is not counted as a directory part so we must resolve the result relative to it.
         if (path.getRoot() != null) {
             checkForSymLink = path.getRoot().resolve(checkForSymLink);
         }
```
Comment (lgoldstein) [inline sshd-common/src/main/java/org/apache/sshd/common/util/io/IoUtils.java]: D.R.Y. 
```java
Path root  = path.getRoot();
if (root != null) {
     checkForSymlink = root.resolve(checkForSymLink);
}
```
Not only it is more efficient, but it is more readable, easier to debug and less likely to create a vulnerability (I can show you how such repeated calls can be leveraged to create security vulnerabilities)

File: sshd-common/src/test/java/org/apache/sshd/common/file/root/RootedFileSystemProviderTest.java [hunk 2]
```diff
@@ -130,9 +131,27 @@ public static void afterClass() {
     }
 
     private Path getTargetFolderOnHostFs(Path targetFolder) {
-        return this.hostFilesystem.getSeparator().equals("\\")
-                ? this.hostFilesystem.getPath("C:", targetFolder.toString())
-                : this.hostFilesystem.getPath(targetFolder.toString());
+        // need to reroot a file for a unix file system on windows
+        if (this.hostFilesystem.getSeparator().equals("/") && targetFolder.getRoot() != null
```
Comment (lgoldstein) [inline sshd-common/src/test/java/org/apache/sshd/common/file/root/RootedFileSystemProviderTest.java]: Please avoid D.R.Y. calls - e.g.,  `hostFilesystem.getSeparator(), targetFolder.getRoot()`, etc. - see previous comment regarding readability, efficiency, debugging and code safety/robustness. I am well aware that this is test code, but we do not use different standards for test vs "main" code (also bear in mind that code that it currently in test may become "main" - in which case, we do not want to re-write it...)

**Review by lgoldstein — CHANGES_REQUESTED** [review]

File: sshd-common/src/test/java/org/apache/sshd/common/file/root/RootedFileSystemProviderTest.java [hunk 3]
```diff
@@ -155,10 +174,15 @@ public void testMkdir() throws IOException {
...
 
-    @Test(expected = NoSuchFileException.class)
-    public void testMkdirInvalid() throws IOException {
-        Path parent = fileHelper.createDirectory(fileSystem.getPath(DOESNT_EXIST + getCurrentTestName()));
-        fail(String.format("Unexpected success in creating directory %s", parent.toString()));
+    @Test
+    public void testMkdirInvalid() {
+        if (isTestingRealWindowsHostFS) {
+            return;
+        }
```
Comment (tomaswolf) [inline sshd-common/src/test/java/org/apache/sshd/common/file/root/RootedFileSystemProviderTest.java]: Better than just returning would be `Assume.assumeFalse(isTestingRealWindowsHostFS);`.

But could you explain why the tests have to be skipped on a real Windows FS?

File: sshd-common/src/test/java/org/apache/sshd/common/file/root/RootedFileSystemProviderTest.java [hunk 4]
```diff
@@ -195,10 +219,15 @@ public void testChdir() throws IOException {
         }
     }
 
-    @Test(expected = NoSuchFileException.class)
+    @Test
     public void testChdirInvalid() throws IOException {
-        Path chdir = fileHelper.createDirectory(fileSystem.getPath(DOESNT_EXIST + getCurrentTestName()));
-        fail(String.format("Unexpected success in changing directory %s", chdir.toString()));
+        if (isTestingRealWindowsHostFS) {
```
Comment (tomaswolf) [inline sshd-common/src/test/java/org/apache/sshd/common/file/root/RootedFileSystemProviderTest.java]: This looks identical to `testMkdirInvalid()` above...

File: sshd-common/src/main/java/org/apache/sshd/common/util/io/IoUtils.java [hunk 5]
```diff
@@ -426,29 +426,40 @@ public static Boolean checkFileExistsAnySymlinks(Path path, boolean neverFollowS
...
                     if (basicFileAttributes.isSymbolicLink()) {
                         return false;
                     }
                 }
             }
-            return Boolean.TRUE;
+            return true;
         } catch (NoSuchFileException e) {
-            return Boolean.FALSE;
+            return false;
```
Comment (tomaswolf) [inline sshd-common/src/main/java/org/apache/sshd/common/util/io/IoUtils.java:440]: I have a problem with the `Boolean` return type. Using a Boolean for a tri-state return is confusing. Also, the javadoc would need to spell out what a `null` return means. Can't this just return `boolean`,  and just return false on IOE? As far as I see, callers map this to an AccessDeniedException anyway.

Comment (pyckle) [reply sshd-common/src/main/java/org/apache/sshd/common/util/io/IoUtils.java:440]: Agreed that this is unclear. I made this API to retrofit `checkFileExists` in many places, so I wanted to make the API similar, so it could be a drop in replacement.

Really, the return code should be  an enum `EXISTS, DOESNT_EXIST, ILLEGAL_SYMLINK_FOUND, UNKNOWN`, for better error codes, but this adds changing logic in multiple places, which I'd prefer to avoid doing right now - this is a large enough PR already..

File: sshd-common/src/main/java/org/apache/sshd/common/util/io/IoUtils.java (hunk 1 above)
Comment (pyckle) [reply sshd-common/src/main/java/org/apache/sshd/common/util/io/IoUtils.java]: > Not only it is more efficient

In any modern JVM on Unix (Linux, MacOS, etc.), this won't be more efficient, as the JVM will optimize this call out. There's a possibility it may be slightly faster on JimFs and Windows, but certainly not enough to justify mentioning performance here.

> is more readable, easier to debug...

This is a matter of style. I'll change it to what you prefer.

> less likely to create a vulnerability

This is not relevant for functions on immutable objects like Path and String.

File: sshd-common/src/test/java/org/apache/sshd/common/file/root/RootedFileSystemProviderTest.java (hunk 3 above)
Comment (pyckle) [reply sshd-common/src/test/java/org/apache/sshd/common/file/root/RootedFileSystemProviderTest.java]: Added `Assume.assumeFalse()` with an error message. I incorrectly assumed that `if(){return;}` would be better as it doesn't show up as a skipped test.

File: sshd-common/src/test/java/org/apache/sshd/common/file/root/RootedFileSystemProviderTest.java (hunk 4 above)
Comment (pyckle) [reply sshd-common/src/test/java/org/apache/sshd/common/file/root/RootedFileSystemProviderTest.java]: Good catch!

It appears that this bug is in master:
https://github.com/apache/mina-sshd/blob/master/sshd-common/src/test/java/org/apache/sshd/common/file/root/RootedFileSystemProviderTest.java#L95
https://github.com/apache/mina-sshd/blob/master/sshd-common/src/test/java/org/apache/sshd/common/file/root/RootedFileSystemProviderTest.java#L130

Would you like me to delete this test?

File: sshd-common/src/main/java/org/apache/sshd/common/util/io/IoUtils.java (hunk 1 above)
Comment (lgoldstein) [reply sshd-common/src/main/java/org/apache/sshd/common/util/io/IoUtils.java]: > In any modern JVM on Unix (Linux, MacOS, etc.), this won't be more efficient, as the JVM will optimize this call out.

Inaccurate  - in order for this to be optimized there would have to be **lots** of calls to the code, and even then it might not be optimized since the JVM cannot be sure that the return value of *getRoot* is the same, so it would not be safe to optimize.

> less likely to create a vulnerability
>> This is not relevant for functions on immutable objects like Path and String.

*Path* is definitely **not** immutable - it is *recommended* to be implemented as such, but it is not necessarily so.

Comment (pyckle) [reply sshd-common/src/main/java/org/apache/sshd/common/util/io/IoUtils.java]: > in order for this to be optimized there would have to be lots of calls to the code

If there aren't many calls to the code, its performance is irrelevant. This is why the Oracle JVM JIT compiler engine is called hotspot - it optimizes only the spots that matter.

> the JVM cannot be sure that the return value of getRoot is the same

Read the implementation of [UnixPath](https://github.com/AdoptOpenJDK/openjdk-jdk11/blob/master/src/java.base/unix/classes/sun/nio/fs/UnixPath.java). Hotspot is smart enough to see the code path in `getRoot()` does not for the same Path object, and that it always returns the same root from [UnixFileSystem](https://github.com/AdoptOpenJDK/openjdk-jdk11/blob/master/src/java.base/unix/classes/sun/nio/fs/UnixFileSystem.java#L92).

In short, there's no practical performance concern. At best, in theory, this might save a few microseconds (more likely nanoseconds) on application startup. And in the words of Donald Knuth, "Premature optimization is the root of all evil." Mentioning performance in irrelevant places is bad practice.

> Path is definitely not immutable

If you have a mutable Path implementation, you probably shouldn't be using standard libraries with it. And you definitely shouldn't be mutating it while passing it into a library function.

In this case, a `Path` implementation that breaks this code is a theoretical exercise. It would require `getRoot()` to mutate its state, or another thread to cause a race condition in `getRoot()`.

File: sshd-common/src/test/java/org/apache/sshd/common/file/root/RootedFileSystemProviderTest.java (hunk 4 above)
Comment (tomaswolf) [reply sshd-common/src/test/java/org/apache/sshd/common/file/root/RootedFileSystemProviderTest.java]: Yes please. Makes no sense.

Comment (pyckle) [reply sshd-common/src/test/java/org/apache/sshd/common/file/root/RootedFileSystemProviderTest.java]: Deleted.

File: sshd-common/src/main/java/org/apache/sshd/common/util/io/IoUtils.java (hunk 1 above)
Comment (lgoldstein) [reply sshd-common/src/main/java/org/apache/sshd/common/util/io/IoUtils.java]: > Premature optimization is the root of all evil

IMO not applicable here - let's agree to disagree. This, however, does not diminish the readability and especially **debugging** concerns. If someone wants to examine the value returned by *getRoot()* it may require some non-trivial setup of the IDE instead of a simple line breakpoint. If only for this reason (but IMO for all the other reasons) this kind (and 99.999%) of D.R.Y. should be avoided.

Comment (lgoldstein) [reply sshd-common/src/main/java/org/apache/sshd/common/util/io/IoUtils.java]: > Read the implementation of [UnixPath](https://github.com/AdoptOpenJDK/openjdk-jdk11/blob/master/src/java.base/unix/classes/sun/nio/fs/UnixPath.java)

- Who says the *Path* being examined is a *UnixPath* or *WindowsPath* ? BTW, we have an *SftpPath* where *getRoot()* might be "expensive" to invoke (it isn't, but it could be..).
- What guarantee do you have that implementation will not change to something more "expensive" ?

Comment (pyckle) [reply sshd-common/src/main/java/org/apache/sshd/common/util/io/IoUtils.java]: > readability and especially debugging concerns

I don't agree with this either, but as I said, it's a matter of style. Repeating the call saves a line of code and a local variable. It's more concise, and easier for me to read. Modern Java debuggers can evaluate an expression at a breakpoint, so I wouldn't find this any easier to debug. Programmer discretion is required, in these sorts of questions, and we came to different conclusions. I understand the your view and think that the extra variable is also acceptable, but I still prefer the other approach.

> Who says the Path being examined is a UnixPath or WindowsPath ?

It's reasonable to assume that a simple function like `getRoot()` on a `Path` object runs fast. Remember that the Path object is a representation of a file in a file system, and the methods on `Path` (other than `toRealPath()`) are done without IO. This is why they cannot throw `IOException`. These operations thus are fast by their nature, and Hotspot does a great job optimizing them.

>  What guarantee do you have that implementation will not change to something more "expensive" ?

In addition to the above analysis, historically the JVM does not make changes that significantly slow performance. I'm happy to bet on this in places where I think it's cleaner. I would agree with this concern if the function `getRoot()` was non-trivial.
